src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,571 @@
+ *     optional.</li>
+ * </ul>
+ *
+ * @see Stream#gather(Gatherer)
+ * @see Gatherers
+ *
+ * @param <T> the type of input elements to the gatherer operation
+ * @param <A> the potentially mutable state type of the gatherer operation
+ *            (often hidden as an implementation detail)
+ * @param <R> the type of output elements from the gatherer operation
+ * @since 22
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+public interface Gatherer<T, A, R> {
+    /**
+     * A function that produces an instance of the intermediate state used for
+     * this gathering operation.
+     *
+     * @implSpec The implementation in this interface returns
+     *           {@link #defaultInitializer()}.
+     *
+     * @return A function that produces an instance of the intermediate state
+     * used for this gathering operation
+     */
+    default Supplier<A> initializer() {
+        return defaultInitializer();
+    };
+
+    /**
+     * A function which integrates provided elements, potentially using
+     * the provided intermediate state, optionally producing output to the
+     * provided {@link Downstream}.
+     *
+     * @return a function which integrates provided elements, potentially using
+     *         the provided state, optionally producing output to the provided
+     *         Downstream
+     */
+    Integrator<A, T, R> integrator();
+
+    /**
+     * A function which accepts two intermediate states and combines them into
+     * one.
+     *
+     * @implSpec The implementation in this interface returns
+     *           {@link #defaultCombiner()}.
+     *
+     * @return a function which accepts two intermediate states and combines
+     *         them into one
+     */
+    default BinaryOperator<A> combiner() {
+        return defaultCombiner();
+    }
+
+    /**
+     * A function which accepts the final intermediate state
+     * and a {@link Downstream} object, allowing to perform a final action at
+     * the end of input elements.
+     *
+     * @implSpec The implementation in this interface returns
+     *           {@link #defaultFinisher()}.
+     *
+     * @return a function which transforms the intermediate result to the final
+     *         result(s) which are then passed on to the provided Downstream
+     */
+    default BiConsumer<A, Downstream<? super R>> finisher() {
+        return defaultFinisher();
+    }
+
+    /**
+     * Returns a composed Gatherer which connects the output of this Gatherer
+     * to the input of that Gatherer.
+     *
+     * @param that the other gatherer
+     * @param <RR> The type of output of that Gatherer
+     * @throws NullPointerException if the argument is null
+     * @return returns a composed Gatherer which connects the output of this
+     *         Gatherer as input that Gatherer
+     */
+    default <RR> Gatherer<T, ?, RR> andThen(Gatherer<? super R, ?, ? extends RR> that) {
```

Comment:
@amaembo @forax Removing the extra type parameter for the RHS state type turned out to be remarkably trivial, so thanks for the idea!